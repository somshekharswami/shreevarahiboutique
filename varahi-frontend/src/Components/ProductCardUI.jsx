import React from "react";
import { Link, useNavigate } from "react-router-dom";
import { useSelector } from "react-redux";
import { motion } from "framer-motion";
import { optimizeCloudinaryUrl } from "../utils/cloudinary";

export default function ProductCardUI({
  pageTitle,
  pageDesc,
  notFoundtitle,
  notFoundDesc,
  products = [],
}) {
  const navigate = useNavigate();
  const cartItems = useSelector((state) => state.cart.cartItems);

  const getQuantity = (_id) => {
    const item = cartItems.find((p) => p._id === _id);
    return item ? item.quantity : 0;
  };

  if (!products || products.length === 0) {
    return (
      <div className="text-center py-16">
        <h1
          className="text-3xl font-serif text-gray-700 mb-4"
          aria-label="notfoundtitle"
        >
          {notFoundtitle || "No Products Found"}
        </h1>
        <p
          className="text-md text-gray-600 font-light"
          aria-label="notfounddesc"
        >
          {notFoundDesc ||
            "Please check back later or explore other categories."}
        </p>
      </div>
    );
  }

  return (
    <>
      <section className="min-h-screen bg-gradient-to-br from-pink-100 to-purple-100 py-12 px-4 sm:px-10">
        <div className="max-w-7xl mx-auto">
          <div className="text-center mb-10">
            <h1 className="text-3xl md:text-5xl font-serif font-semibold text-[#b11257] text-center mt-1">
              {pageTitle || "Shree Varahi Boutique"}
            </h1>
            <p className="text-md sm:text-lg text-gray-700 font-light font-serif max-w-2xl mx-auto leading-relaxed">
              {pageDesc ||
                "Explore our handpicked collection of ethnic fashion. Elegance and tradition woven into every thread."}
            </p>
          </div>

          <div className="grid grid-cols-2 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 xl:grid-cols-5 gap-6">
            {products.map((product) => (
              <motion.div
                key={product._id}
                className="bg-white rounded-2xl shadow-md overflow-hidden group hover:shadow-xl transition-shadow duration-300"
                whileHover={{ scale: 1.02 }}
              >
                <div className="relative overflow-hidden bg-gray-50">
                  <img
                    src={optimizeCloudinaryUrl(product.imageUrl, 256)}
                    alt={product.title}
                    loading="lazy"
                    onError={(e) => {
                      e.target.onerror = null;
                      e.target.src = "/fallback.jpg";
                    }}
                    className="w-full h-48 sm:h-56 object-contain group-hover:scale-105 transition-transform duration-300 p-2"
                  />
                </div>

                <div className="p-4">
                  <h2 className="text-base sm:text-lg font-semibold text-gray-900 mb-2 line-clamp-2 group-hover:text-pink-600 transition-colors font-serif">
                    {product.title}
                  </h2>
                  <p className="text-gray-700 text-sm sm:text-base font-normal font-serif line-clamp-2 mb-2">
                    {product.description}
                  </p>
                  <div className="flex items-center justify-between mb-3">
                    <div className="text-xl font-sans">
                      {/* Original Price - Strikethrough */}
                      <span className="line-through text-pink-700 mr-2">
                        ₹
                        {product.singlePrice?.originalPrice
                          ? product.singlePrice.originalPrice
                          : product.sizes?.length > 0
                          ? Math.min(
                              ...product.sizes.map((s) => s.originalPrice)
                            )
                          : "N/A"}
                      </span>

                      {/* Discounted Price - Highlighted */}
                      <span className="font-bold text-green-700">
                        ₹
                        {product.singlePrice?.discountedPrice
                          ? product.singlePrice.discountedPrice
                          : product.sizes?.length > 0
                          ? Math.min(
                              ...product.sizes.map((s) => s.discountedPrice)
                            )
                          : "N/A"}
                      </span>
                    </div>
                  </div>

                  <button
                    onClick={(e) => {
                      e.stopPropagation();
                      navigate(`/product/${product._id}`);
                    }}
                    className="w-full font-sans px-4 py-2 mt-2 bg-pink-600 hover:bg-pink-700 text-white rounded-xl shadow-md transition-all duration-200"
                  >
                    View Item
                  </button>
                </div>
              </motion.div>
            ))}
          </div>
        </div>
      </section>
    </>
  );
}
