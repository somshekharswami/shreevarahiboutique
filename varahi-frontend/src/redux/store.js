import { configureStore } from "@reduxjs/toolkit";
import cartReducer from "./slices/cartSlice";
import authReducer from "./slices/authSlice"; 

// Load cart from localStorage
const loadCart = () => {
  try {
    const saved = localStorage.getItem("cartState");
    return saved ? JSON.parse(saved) : undefined;
  } catch {
    return undefined;
  }
};

// Save cart to localStorage
const saveCart = (cartSlice) => {
  try {
    localStorage.setItem("cartState", JSON.stringify(cartSlice));
  } catch {
    /* ignore write errors */
  }
};

const preloadedState = loadCart() ? { cart: loadCart() } : undefined;

export const store = configureStore({
  reducer: {
    cart: cartReducer,
    auth: authReducer,
  },
  preloadedState,
});

store.subscribe(() => {
  saveCart(store.getState().cart); // persist only cart
});

export default store;
