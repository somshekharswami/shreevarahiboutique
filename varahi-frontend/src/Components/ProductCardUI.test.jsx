import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ProductCardUI from "./ProductCardUI";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import store from "../redux/store";

describe("Testing the ProductCard Component", () => {
  test("Test Case 1: Renders fallback UI when no products", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ProductCardUI
            products={[]}
            notFoundtitle="No Products Found"
            notFoundDesc="Try searching again"
            pageTitle="Test Title"
            pageDesc="Test Desc"
          />
        </BrowserRouter>
      </Provider>
    );

    const notFoundtitle = screen.getByLabelText(/notfoundtitle/i);
    const notFounddesc = screen.getByLabelText(/notfounddesc/i);

    expect(notFoundtitle).toBeInTheDocument();
    expect(notFounddesc).toBeInTheDocument();
    expect(notFoundtitle).toHaveTextContent("No Products Found");
    expect(notFounddesc).toHaveTextContent("Try searching again");
  });

  test(" Test Case 2: Renders correct number of products.", () => {
    const mockProducts = [
      {
        _id: "1",
        title: "Product 1",
        description: "Description 1",
        imageUrl: "/image1.jpg",
        category: "Dupatta",
        singlePrice: { discountedPrice: 500 },
      },
      {
        _id: "2",
        title: "Product 2",
        description: "Description 2",
        imageUrl: "/image2.jpg",
        category: "Dress",
        singlePrice: { discountedPrice: 800 },
      },
    ];
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ProductCardUI
            products={mockProducts}
            pageTitle="Our Products"
            pageDesc="Best styles available"
            notFoundtitle="No Products"
            notFoundDesc="Try again"
          />
        </BrowserRouter>
      </Provider>
    );
    const productTitles = screen.getAllByRole("heading", { level: 2 }); // to find all product <h2> tags.
    expect(productTitles.length).toBe(mockProducts.length);
    // displays correct product title
    expect(screen.getByText("Product 1")).toBeInTheDocument();
    expect(screen.getByText("Product 2")).toBeInTheDocument();
  });

  test("Displays the correct price", () => {
    const mockProducts = [
      {
        _id: "1",
        title: "Product With Single Price",
        description: "heyy",
        imageUrl: "https://via.placeholder.com/150",
        category: "Dress",
        singlePrice: { discountedPrice: 999 },
      },
      {
        _id: "2",
        title: "Product With Sizes",
        description: "descc",
        imageUrl: "https://via.placeholder.com/150",
        category: "Kurti",
        sizes: [
          { label: "S", discountedPrice: 500 },
          { label: "M", discountedPrice: 450 },
        ],
      },
    ];
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ProductCardUI
            products={mockProducts}
            pageTitle="Price Check"
            pageDesc="Testing price logic"
            notFoundtitle="No Products"
            notFoundDesc="Try again"
          />
        </BrowserRouter>
      </Provider>
    );
    expect(screen.getByText("₹999")).toBeInTheDocument(); // from singlePrice
    expect(screen.getByText("₹450")).toBeInTheDocument(); // min from sizes
  });
  test("Match Snapshots", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <ProductCardUI
            pageTitle="Snapshot Title"
            pageDesc="This is a snapshot test"
            notFoundtitle="No Products Found"
            notFoundDesc="Try again later"
            products={[]}
          />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment).toMatchSnapshot();
  });
});
