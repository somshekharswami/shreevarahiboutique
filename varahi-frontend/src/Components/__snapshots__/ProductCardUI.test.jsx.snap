// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Testing the ProductCard Component Match Snapshots 1`] = `[Function]`;
